README.md

```diff
@@ -11,572 +11,528 @@
-in the stream concurrently when used with appropriate stream type combinator
-(e.g. `fromAsync`, `fromAhead` or `fromParallel`).
+``` haskell
+import qualified Streamly.Data.Fold as Fold
+import qualified Streamly.Internal.FileSystem.File as File
+import qualified Streamly.Prelude as Stream
+
+wcb :: String -> IO Int
+wcb file =
+    File.toBytes file        -- SerialT IO Word8
+  & Stream.fold Fold.length  -- IO Int
+```
 
-The following code finishes in 3 seconds (6 seconds when serial), note the
-order of elements in the resulting output, the outputs are consumed as soon as
-each action is finished (asyncly):
+### Count lines (wc -l)
+
+Count lines in a file.
 
 ``` haskell
-> let p n = threadDelay (n * 1000000) >> return n
-> S.toList $ S.fromAsync $ p 3 |: p 2 |: p 1 |: S.nil
-[1,2,3]
+-- ASCII character 10 is newline
+countl :: Int -> Word8 -> Int
+countl n ch = if ch == 10 then n + 1 else n
+
+-- The fold accepts a stream of `Word8` and returns a line count (`Int`).
+nlines :: Monad m => Fold m Word8 Int
+nlines = Fold.foldl' countl 0
+
+wcl :: String -> IO Int
+wcl file =
+    File.toBytes file  -- SerialT IO Word8
+  & Stream.fold nlines -- IO Int
 ```
 
-Use `fromAhead` if you want speculative concurrency i.e. execute the actions in
-the stream concurrently but consume the results in the specified order:
+### Count words (wc -w)
+
+Count words in a file.
 
 ``` haskell
-> S.toList $ S.fromAhead $ p 3 |: p 2 |: p 1 |: S.nil
-[3,2,1]
+countw :: (Int, Bool) -> Word8 -> (Int, Bool)
+countw (n, wasSpace) ch =
+    if isSpace $ chr $ fromIntegral ch
+    then (n, True)
+    else (if wasSpace then n + 1 else n, False)
+
+-- The fold accepts a stream of `Word8` and returns a word count (`Int`)
+nwords :: Monad m => Fold m Word8 Int
+nwords = fst <$> Fold.foldl' countw (0, True)
+
+wcw :: String -> IO Int
+wcw file =
+    File.toBytes file   -- SerialT IO Word8
+  & Stream.fold nwords  -- IO Int
 ```
 
-Monadic stream generation functions e.g. `unfoldrM`, `replicateM`, `repeatM`,
-`iterateM` and `fromFoldableM` etc. can work concurrently.
+### Count bytes, words and lines
 
-The following finishes in 10 seconds (100 seconds when serial):
+We can compose the three folds together into a single fold using `Tee`
+to do all the three things at once. The applicative instance of `Tee`
+distributes the input to all the folds and combines the outputs using the
+supplied function.
 
 ``` haskell
-S.drain $ S.fromAsync $ S.replicateM 10 $ p 10
+import qualified Streamly.Internal.Data.Fold.Tee as Tee
+
+-- The fold accepts a stream of `Word8` and returns the three counts
+countAll :: Fold IO Word8 (Int, Int, Int)
+countAll = Tee.toFold $ (,,) <$> Tee Fold.length <*> Tee nlines <*> Tee nwords
+
+wc :: String -> IO (Int, Int, Int)
+wc file =
+    File.toBytes file    -- SerialT IO Word8
+  & Stream.fold countAll -- IO (Int, Int, Int)
 ```
 
-## Concurrency Auto Scaling
+### Observations
```

Comment:
I don't know how I feel about observations. Is there a better way to paraphrase this at the end?

Reply:
We can remove it or just add a line in the previous section itself.